### PR TITLE
Rewrite CA2213 (DisposableFieldsShouldBeDisposed) analyzer to be a SymbolStart/End analyzer

### PR DESCRIPTION
**Pull request by mavasani**
Work towards reducing CompilationEnd analyzers in the repo: #2579

### CODE REVIEW

**Comment by mavasani** [comment]
FYI: Much easier to review with https://github.com/dotnet/roslyn-analyzers/pull/3235/files?w=1

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs [hunk 1]
```diff
@@ -41,172 +41,217 @@ public override void Initialize(AnalysisContext context)
...
-                compilationContext.RegisterOperationAction(_ => hasErrors = true, OperationKind.Invalid);
+            // Instance fields initialized in constructor/method body with a locally created disposable object must be disposed.
+            symbolStartContext.RegisterOperationBlockStartAction(OnOperationBlockStart);
 
-                // Disposable fields with initializer at declaration must be disposed.
-                compilationContext.RegisterOperationAction(operationContext =>
+            // Report diagnostics at symbol end.
+            symbolStartContext.RegisterSymbolEndAction(OnSymbolEnd);
+
+            return;
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs]: Do you need this return?

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs [hunk 2]
```diff
@@ -41,172 +41,217 @@ public override void Initialize(AnalysisContext context)
...
-                        }
-                    }
-                },
-                OperationKind.FieldInitializer);
+                return false;
+            }
+
+            void OnFieldInitializer(OperationAnalysisContext operationContext)
+            {
+                RoslynDebug.Assert(disposeAnalysisHelper != null);
```
Comment (Evangelink) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs]: Is there any guideline on when to use `RoslynDebug` or `Debug`?

Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs]: `RoslynDebug.Assert` is something that is nullable annotated, so we use it only when we get a false nullable warning from the C# compiler. Other alternative is to use the `!` suppression operator at use sites for this variable.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs (hunk 1 above)
Comment (mavasani) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs]: It is generally recommended to have an explicit return statement separating the core logic and local functions, which are primarily implementation details. We do so even in roslyn repo.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs (hunk 2 above)
Comment (Evangelink) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DisposableFieldsShouldBeDisposed.cs]: Thanks

**Pull request merged**